File: mpfreader.rb
require 'socket'                # Get sockets from stdlib

server = TCPServer.open(2000)   # Socket to listen on port 2000
loop do                          # Servers run forever
  Thread.start(server.accept) do |client|
    puts "Accepted the connection. Hello!"
    loop do
      resp = client.recv( 1024 )          # Read complete response
      print "=> "
      resp.bytes { |c| print c.to_s(16), ' ' }
      puts
      
      # Format a ACK MPF packet...
      ackPacket = [ 0x02,         # start of transmission
                    0x06,         # positive acknowledgement
                    1,            # last valid sequence received
                    0x03          # end of transmission
                    ].pack("cccc")

      # Compute LRC
      lrc = ackPacket.byteslice(1, ackPacket.bytesize-1).bytes.inject { |a,b| a ^ b }

      # Append LRC byte to the MPF packet
      ackPacket << lrc
      print "<= "
      ackPacket.bytes { |c| print c.to_s(16), ' ' }
      puts

      # Send the MPF packet
      client.send( ackPacket, 0 )
    end
  end
end

File: mpfreader.js
var net = require('net'),
    events = require('events'),
    mpf = require('./mpf');


var server = net.createServer(function (stream) {
  console.log("new client...");

  // window size
  var windowsize = 7;

  var mpfClient = mpf.createClient(stream);
  mpfClient.on('packet', function (packet) {
    console.log(packet);  
    
    if (--windowsize == 0) {
      // send ack
      var seqno = packet.SeqNo;
      if (seqno) {
        var type = mpf.MPF_PACKET_TYPE_NAK;
        packet = { PacketType: type, SeqNo: seqno };
        console.log("sending packet: " + JSON.stringify(packet));
        mpfClient.sendPacket(packet);
        windowsize = 7;
      }
    }
  });  
}).listen(2000, "127.0.0.1", function() {
    console.log("waiting for connections on port 2000...");
  });
